lib/utils.pm

```diff
@@ -250,6 +251,22 @@ sub get_netboot_mirror {
     return get_var('MIRROR_' . uc($m_protocol));
 }
 
+# repeat zypper_call without failure and record_info
+# avoid failures due to busy server or network
+
+sub zypper_call_retry {
+    my ($command, %args, $repeats) = @_;
+    $repeats //= 5;
+    for (1 .. $repeats) {
+        eval { zypper_call("$command", %args); };
+        last unless ($@);
+        record_info 'zypper failed', 'retry';
+        diag "zypper up failed: $@";
+        diag "Maybe the network or server is busy. Retry: $_ of $repeats";
+    }
+    die "zypper up failed (with retries)" if $@;
+}
```

Comment:
I doubt this is a good idea because we do not have good decision rules where to use zypper_call and where zypper_call_retry. If this really fixes it shouldn't it be included within zypper itself?

Reply:
There are three options abort/retry/ignore when running zypper in interactive mode, but you can not choose how to behave in non-interactive mode, if yes then I could not find it.

Reply:
It's not about some decision, tests are failing because zypper in non-interactive move will abort refresh or other action. If zypper itself is not tested then I think it's waste of hardware power for failed tests due to stupid zypper abort.

Reply:
I have this problem also on real server too..
```
uv300x:~/:[0]# zypper ref
Repository 'SLES12-SP3-Pool' is up to date.                                                                           
Download (curl) error for 'https://updates.suse.com/SUSE/Updates/SLE-SERVER/12-SP3/x86_64/update/media.1/media?w-CD05vId0fNNOlEQxjAqGVxdpu5jT-3ltccSyvynGT5KjeDaETHMRXE8zyLsb-Ut4XWFH1z_UlcF7t1KXxM2So4RXonMvw29rpRVJ1e9lD6v0H5sghAICAuu8B7eCr-11sx7YZKecBU2N8':
Error code: HTTP response: 502
Error message: The requested URL returned error: 502 Bad Gateway

Abort, retry, ignore? [a/r/i/...? shows all options] (a): r
Abort, retry, ignore? [a/r/i/...? shows all options] (a): r
Repository 'SLES12-SP3-Updates' is up to date.                                                                        
Retrieving repository 'Providing openQA dependencies (SLE_12_SP3)' metadata ....................................[done]
Building repository 'Providing openQA dependencies (SLE_12_SP3)' cache .........................................[done]
All repositories have been refreshed.
```

I understand this solution as temporary fix for scc issues. Maybe contact scc team to make sure they know about it.

Reply:
this is not scc - updates.suse.com is the CDN and they have no control over it :(

But yes, they are aware of the problem - I'm not aware that it's tracked somewhere

Reply:
this is what I am saying "stupid zypper abort". For me you are trying to workaround a product problem.

@coolo not tracked like https://bugzilla.suse.com/show_bug.cgi?id=1042733 or https://bugzilla.suse.com/show_bug.cgi?id=992302 ?

Reply:
Exactly, not tracked like these :)

Reply:
To not pollute utils, should I put zypper_call_retry into test where I need it ?

Reply:
you need this in tons of tests - at least as long as the CDN is as flaky

Reply:
do we have better workaround ?